spec/kinesis_client/kinesis_client_spec.rb

```diff
@@ -162,8 +221,59 @@ def error(*args)
       @kinesis_client << '5'
       @kinesis_client.push_records
     end
+
+    it "should log records that failed to enter the kinesis stream" do
+      
+    end 
   end
 
+  describe "#filter_failures", only:true do
+  before(:each) do
+    @mock_failed_response = double
+    @mock_failed_record = double
+    @mock_success_record = double
+    allow(@mock_failed_record).to receive(:error_message).and_return("error")
+    allow(@mock_failed_record).to receive(:key?).with(:error_message).and_return(true)
+    allow(@mock_success_record).to receive(:key?).with(:error_message).and_return(false)
+    allow(@mock_failed_response).to receive(:failed_record_count).and_return(1)
+    allow(@mock_failed_response).to receive(:records)
+      .and_return([@mock_success_record, @mock_failed_record])
+
+  end
+  it "should return records that failed to enter the kinesis stream" do
+      @kinesis_client << '1'
+      @kinesis_client << '2'
+
+    expect(@kinesis_client.filter_failures(@mock_failed_response)).to eql([{record_data:"2",error_message:"error"}])
+  end 
+
+  it "should return an array that can be logged in #push_batch" do
+  allow(@mock_client).to receive(:put_records).with({
+        records: [
+          {
+            data: "encoded 4",
+            partition_key: "hashed"
+          },
+          {
+            data: "encoded 5",
+            partition_key: "hashed"
+          }
+        ],
+        stream_name: 'fake-stream'
+      }).and_return(@mock_failed_response)
+
+    message = {:failures=>1,:failures_data=>{:record_data=>'5',error_message:"error"}}.to_json
+    expect(@kinesis_client).to receive(:push_batch).with([{:data=>"encoded 4", :partition_key=>"hashed"},{:data=>"encoded 5", :partition_key=>"hashed"}]).and_return({
```

Comment:
We don't want to expect the object under test to receive a method. That is an aspect of internal implementation, rather than externally observable behavior. 

How else to test that this is doing what it is supposed to? To me, it looks like that is currently impossible, which is a very useful hint that something is not right with the code. See above where I mention that we are just swallowing the return value of `push_batch`: the fact that we are just swallowing the return value instead of doing anything with it is exactly what makes this impossible to test. So this is a case where the requirement to test a certain way helps to catch an error.

Once we write the code to log errors, we can change this test to check that the logger is receiving the log message.